DataFormats/FEDRawData/src/FEDNumbering.cc

```diff
@@ -3,180 +3,357 @@
+      {
+        in[i] = true;
+      }
+    for(i=FEDNumbering::MINLUMISCALERSFEDID; i<=FEDNumbering::MAXLUMISCALERSFEDID; i++)
+      {
+        in[i] = true;
+      }
+    for(i=FEDNumbering::MINCSCFEDID; i<=FEDNumbering::MAXCSCFEDID; i++)
+      {
+        in[i] = true;
+      }
+    for(i=FEDNumbering::MINCSCTFFEDID; i<=FEDNumbering::MAXCSCTFFEDID; i++)
+      {
+        in[i] = true;
+      }
+    for(i=FEDNumbering::MINDTFEDID; i<=FEDNumbering::MAXDTFEDID; i++)
+      {
+        in[i] = true;
+      }
+    for(i=FEDNumbering::MINDTTFFEDID; i<=FEDNumbering::MAXDTTFFEDID; i++)
+      {
+        in[i] = true;
+      }
+    for(i=FEDNumbering::MINRPCFEDID; i<=FEDNumbering::MAXRPCFEDID; i++)
+      {
+        in[i] = true;
+      }
+    for(i=FEDNumbering::MINTriggerGTPFEDID; i<=FEDNumbering::MAXTriggerGTPFEDID; i++)
+      {
+        in[i] = true;
+      }
+    for(i=FEDNumbering::MINTriggerEGTPFEDID; i<=FEDNumbering::MAXTriggerEGTPFEDID; i++)
+      {
+        in[i] = true;
+      }
+    for(i=FEDNumbering::MINTriggerGCTFEDID; i<=FEDNumbering::MAXTriggerGCTFEDID; i++)
+      {
+        in[i] = true;
+      }
+    for(i=FEDNumbering::MINTriggerLTCFEDID; i<=FEDNumbering::MAXTriggerLTCFEDID; i++)
+      {
+        in[i] = true;
+      }
+    for(i=FEDNumbering::MINTriggerLTCmtccFEDID; i<=FEDNumbering::MAXTriggerLTCmtccFEDID; i++)
+      {
+        in[i] = true;
+      }
+    for(i=FEDNumbering::MINCSCDDUFEDID; i<=FEDNumbering::MAXCSCDDUFEDID; i++)
+      {
+        in[i] = true;
+      }
+    for(i=FEDNumbering::MINCSCContingencyFEDID; i<=FEDNumbering::MAXCSCContingencyFEDID; i++)
+      {
+        in[i] = true;
+      }
+    for(i=FEDNumbering::MINCSCTFSPFEDID; i<=FEDNumbering::MAXCSCTFSPFEDID; i++)
+      {
+        in[i] = true;
+      }
+    for(i=FEDNumbering::MINDAQeFEDFEDID; i<=FEDNumbering::MAXDAQeFEDFEDID; i++)
+      {
+        in[i] = true;
+      }
+    for(i=FEDNumbering::MINDAQmFEDFEDID; i<=FEDNumbering::MAXDAQmFEDFEDID; i++)
+      {
+        in[i] = true;
+      }
+    for(i=FEDNumbering::MINTCDSuTCAFEDID; i<=FEDNumbering::MAXTCDSuTCAFEDID; i++)
+      {
+        in[i] = true;
+      }
+    for(i=FEDNumbering::MINHCALuTCAFEDID; i<=FEDNumbering::MAXHCALuTCAFEDID; i++)
+      {
+        in[i] = true;
+      }
+    for(i=FEDNumbering::MINSiPixeluTCAFEDID; i<=FEDNumbering::MAXSiPixeluTCAFEDID; i++)
+      {
+        in[i] = true;
+      }
+    for(i=FEDNumbering::MINDTUROSFEDID; i<=FEDNumbering::MAXDTUROSFEDID; i++)
+      {
+        in[i] = true;
+      }
+    for(i=FEDNumbering::MINTriggerUpgradeFEDID; i<=FEDNumbering::MAXTriggerUpgradeFEDID; i++)
+      {
+        in[i] = true;
+      }
+    return in;
+  }
+
+  constexpr std::array<int, FEDNumbering::MAXFEDID+1> initFromIndex() {
+    std::array<int, FEDNumbering::MAXFEDID+1> fromIndex= {{0}};
+    int i = 0;
+    int index = 0;
+    for(i=0; i< FEDNumbering::lastFEDId(); i++)
+      fromIndex[i] = index; //empty
+    ++index;
```

Comment:
can't the ```fromString```  be filled here as well?
... and also all of this code merged together with the definition of ```in ``` in ```initIn```.

Reply:
It can not be made `constexpr`. All the `constexpr` functions are run at compile time with the arrays the created filled at compile time. There is no runtime. One can not fill std::string at compile time.

I added a very extensive test which gets run at each pull request which makes sure the values are correct.

Reply:
> I added a very extensive test which gets run at each pull request which makes sure the values are correct.

this only tests the code that exists now; there is no mechanism to require/fail the unit test if a new min/max pair is added.

Reply:
I wonder if we could get rid of the string handling altogether. The method FEDNumbering::fromDet is not used anywhere in CMSSW according to http://cmslxr.fnal.gov/ident?_i=fromDet

Reply:
this would be much easier.
Thanks for the proposal.

Reply:
Is the consensus for the string handling to be removed? If so, I will make that change.

Reply:
daq and reco are apparently OK with this.

Reply:
IMHO, it can be removed.